text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1018 @@
+support the following scenarios: cost efficiency, advanced data processing, data minimization. New receivers, processors,
+and exporters supporting Apache Arrow natively will be developed. A bidirectional adaptation layer OTLP / OTel Arrow
+will be developed within the collector to continue supporting the existing ecosystem. The following diagram is an
+overview of a collector supporting both OTLP and an end-to-end OTel Arrow pipeline.
+
+![OTel Arrow Collector](img/0156_collector_phase_2.png)
+
+Implementing an end-to-end column-oriented pipeline will provide many benefits such as:
+
+- **Accelerate stream processing**,
+- **Reduce CPU and Memory usage**,
+- **Improve compression ratio end-to-end**,
+- **Access to the Apache Arrow ecosystem** (query engine, parquet support, ...).
+
+## Protocol Details
+
+The protocol specifications are composed of two parts. The first section describes the new gRPC services supporting
+column-oriented telemetry data. The second section presents the mapping between the OTLP entities and their Apache
+Arrow counterpart.
+
+### ArrowStreamService
+
+OTel Arrow defines the columnar encoding of telemetry data and the gRPC-based protocol used to exchange data between
+the client and the server. OTel Arrow is a bi-directional stream oriented protocol leveraging Apache Arrow for the
+encoding of the telemetry data.
+
+OTLP and OTel Arrow protocols can be used together and can use the same TCP port. To do so, in addition to the 3
+existing
+services (`MetricsService`, `LogsService` and `TraceService`), we introduce the service `ArrowStreamService`
+(see [this protobuf specification](#appendix-a---protocol-buffer-definitions) for more details) exposing a single API
+endpoint named `ArrowStream`. This endpoint is based on a bidirectional streaming protocol. The client message is a
+`BatchArrowRecords` stream encoding a batch of Apache Arrow buffers (more specifically [Arrow IPC format](#arrow-ipc-format)).
+The server message side is a `BatchStatus` stream reporting asynchronously the status of each `BatchArrowRecords`
+previously sent. In addition to this endpoint, the OTel Arrow protocol offers three additional services to facilitate
+intricate load-balancing routing rules, tailored to the specific nature of the OTLP entities - namely Metrics, Logs,
+and Traces.
+
+After establishing the underlying transport the client starts sending telemetry data using the `ArrowStream` service.
+The client continuously sends `BatchArrowRecords`'s messages over the opened stream to the server and expects to receive
+continuously `BatchStatus`'s messages from the server as illustrated by the following sequence diagram:
+
+![Sequence diagram](img/0156_OTEL%20-%20ProtocolSeqDiagram.png)
+
+> Multiple streams can be simultaneously opened between a client and a server to increase the maximum achievable
+throughput.
+
+If the client is shutting down (e.g. when the containing process wants to exit) the client will optionally wait until
```

Comment:
I believe the contract of an exporter is that they must drain their queue on exit, so it shouldn’t be optional.

Reply:
See https://github.com/open-telemetry/oteps/pull/171/commits/7fbd2693524f86edc59946141d52ef77028fc717